### PR TITLE
Move active to ServiceEndpoint so that user can decide whether send or not

### PR DESCRIPTION
**Pull request by vicancy**
Move active to ServiceEndpoint so that user can decide whether to send messages or not instead of making it a default behavior of weak connections

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 1]
```diff
@@ -26,6 +21,16 @@ public class ServiceEndpoint
 
         internal int? Port { get; }
 
+        /// <summary>
+        /// When there are server connections connected to the target endpoint for current hub, the endpoint is considered as Online; otherwise, Offline.
+        /// </summary>
+        public bool Online { get; internal set; } = true;
+
+        /// <summary>
+        /// When the target endpoint has client connections of current hub, the endpoint is considered as Active; otherwise, inactive.
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:30]: Active ->  active

File: test/Microsoft.Azure.SignalR.Tests.Common/TestClasses/TestServiceConnection.cs [hunk 2]
```diff
@@ -26,12 +26,15 @@ internal class TestServiceConnection : ServiceConnectionBase
...
+        public TestServiceConnection(ServiceConnectionStatus status = ServiceConnectionStatus.Connected, bool throws = false,
+            ILogger logger = null,
+            IServiceMessageHandler serviceMessageHandler = null
+            ) : base(
             new ServiceProtocol(),
             "serverId",
             Guid.NewGuid().ToString(),
             new HubServiceEndpoint(),
-            null, // TODO replace it with a NullMessageHandler
+            serviceMessageHandler, // TODO replace it with a NullMessageHandler
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Tests.Common/TestClasses/TestServiceConnection.cs:37]: delete the comment?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 3]
```diff
@@ -345,6 +351,33 @@ protected async Task RemoveConnectionAsync(IServiceConnection c, bool migratable
...
+            {
+                _inactiveInfo = (0, null);
+                return true;
+            }
+            else
+            {
+                var info = _inactiveInfo;
+                var last = info.last ?? DateTime.UtcNow;
+                var count = info.count;
+                count++;
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs:366]: not thread safe, could multiple status pings make results inaccurate?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 4]
```diff
@@ -345,6 +351,33 @@ protected async Task RemoveConnectionAsync(IServiceConnection c, bool migratable
...
+                _inactiveInfo = (count, last);
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs:367]: not thread-safe, could multiple status ping near similar time turns out to be one?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 5]
```diff
@@ -345,6 +351,33 @@ protected async Task RemoveConnectionAsync(IServiceConnection c, bool migratable
...
+                
+                // Inactive it only when it checks over 5 times and elapsed for over 10 minutes
+                if (count >= checkWindow && DateTime.UtcNow - last >= checkTimeSpan)
+                {
+                    return false;
+                }
+                else
+                {
+                    return true;
+                }
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs:377]: return !(count >= checkWindow && DateTime.UtcNow - last >= checkTimeSpan)

File: test/Microsoft.Azure.SignalR.Tests/MultiEndpointServiceConnectionContainerTests.cs [hunk 6]
```diff
@@ -33,6 +33,56 @@ public TestEndpointServiceConnectionContainerTests(ITestOutputHelper output) : b
...
+                {
+                    new ServiceEndpoint(ConnectionString1, EndpointType.Primary, "strong"),
+                    new ServiceEndpoint(ConnectionString2, EndpointType.Secondary, "weak"),
+                };
+
+                var sem = new TestServiceEndpointManager(endpoints);
+
+                var router = new TestEndpointRouter();
+
+                var connectionFactory1 = new TestServiceConnectionFactory();
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Tests/MultiEndpointServiceConnectionContainerTests.cs:51]: connectionFactory1 -> connectionFactory ?

File: test/Microsoft.Azure.SignalR.Tests/MultiEndpointServiceConnectionContainerTests.cs [hunk 7]
```diff
@@ -33,6 +33,56 @@ public TestEndpointServiceConnectionContainerTests(ITestOutputHelper output) : b
...
+
+                var hub1 = new MultiEndpointServiceConnectionContainer(connectionFactory1, "hub1", 2, sem, router,
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Tests/MultiEndpointServiceConnectionContainerTests.cs:53]: hub1 -> hub ?

File: test/Microsoft.Azure.SignalR.Tests/MultiEndpointServiceConnectionContainerTests.cs [hunk 8]
```diff
@@ -33,6 +33,56 @@ public TestEndpointServiceConnectionContainerTests(ITestOutputHelper output) : b
...
+                    loggerFactory);
+
+                var connections = connectionFactory1.CreatedConnections.ToArray();
+                Assert.Equal(4, connections.Length);
+
+                var connection1 = connections[0] as TestServiceConnection;
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Tests/MultiEndpointServiceConnectionContainerTests.cs:59]: connection1 => connection ?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 9]
```diff
@@ -26,6 +21,16 @@ public class ServiceEndpoint
 
         internal int? Port { get; }
 
+        /// <summary>
+        /// When there are server connections connected to the target endpoint for current hub, the endpoint is considered as Online; otherwise, Offline.
+        /// </summary>
+        public bool Online { get; internal set; } = true;
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:27]: Can we call it something like `HasServerConnected`? `Online` looks like describing the service is available or not

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 10]
```diff
@@ -26,6 +21,16 @@ public class ServiceEndpoint
...
+
+        /// <summary>
+        /// When the target endpoint has client connections of current hub, the endpoint is considered as Active; otherwise, inactive.
+        /// </summary>
+        public bool IsActive { get; internal set; } = true;
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:32]: Can we call it something like `HasClientConnected`?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs (hunk 4 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs:367]: maybe, but overwrites does not matter

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs (hunk 9 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:27]: It is public and hard to change. and I think Online means the service is available to this app server instance which sounds ok to me.

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:27]: And when all server connections are offline, the service looks unreachable to the server, it is offline.

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs (hunk 10 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:32]: From server's perspective, I think `IsActive` sounds acceptable though, with comments as further explanation.

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs (hunk 4 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs:367]: That will affect the result about count and even about active/inactive check, like inactive may be delayed to know. Is this expected and acceptable?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs (hunk 1 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:30]: intentionally capitalized.

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs (hunk 9 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:27]: add more comments

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs (hunk 10 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:32]: Add more comments

**Pull request merged**